### PR TITLE
(cheevos) include achievement runtime state in save states

### LINKED ISSUE #11894: Softcore achievements issue with save/load states
# First and foremost consider this:
- Only RetroArch bugs should be filed here. Not core bugs or game bugs
- This is not a forum or a help section, this is strictly developer oriented

## Description

This issue described here: https://docs.retroachievements.org/Why-you-shouldn't-use-the-load-state-feature/

Earlier it was in all Retroachievements-supported emulators, now it happened only in Retroarch.

### Expected behavior

Achievements should triggered properly, so internal achievement information should be in save state too.

### Actual behavior

> 
> Example 1: Beat the game without passwordslink
> 
> Imagine a game with 5 stages and with the password feature.
> 
> Probably the conditions to trigger the "Beat the game without passwords" are (in a real cheevo there will be more conditions, but for this explanation let's consider only these ones):
> 
>     be at stage 1 (and memorize this event as something that happened)
>     be at stage 2 (memorize it)
>     be at stage 3 (memorize it)
>     be at stage 4 (memorize it)
>     be at stage 5 (memorize it)
>     finish stage 5
> 
> So to get this achievement you have to pass through all stages and then beat the game.
> 
> When you are at the stage 1, the emulator memorizes it, same thing for all other stages. And when you finish the stage 5 (beat the game) you get the achievement.
> 
> Now imagine this scenario:
> 
>     a softcore player started the game and played stages 1, 2 and 3
>     he needed to stop playing when at stage 4 and then he saved a state to continue later
>     he closed the emulator
>     later he could play again, started the game and loaded the state he saved
>     played stages 4, 5 and beated the game
> 
> In this scenario the achievement won't trigger.
> 
> After closing the emulator it "forgets" that you passed through stages 1, 2 and 3. When you run the emulator/game again and load the state being on stage 4, the emulator won't detect that you passed through the previous stages. Then the achievement won't trigger.
> Example 2: Defeat boss X without taking damagelink
> 
> Most likely the conditions to be met in a frame to make this achievement trigger are:
> 
>     entered the boss room (and memorize this event as something that happened).
>     if take any damage, forget it entered the boss room.
>     beat the boss.
> 
> In other words: to get this achievement the emulator must "know" that you entered the boss room, didn't get any damage, and beated the boss.
> 
> The important detail to pay attention here: when you get any damage, the emulator "forgets" that you entered the boss room. Then even if you load a state to a point where you didn't get any damage but already entered the room, the emulator won't detect that you entered the room. And then the achievement won't trigger.

### Steps to reproduce the bug

1. [First step]
2. [Second step]
3. [and so on...]

### Bisect Results

[Try to bisect and tell us when this started happening]

### Version/Commit
You can find this information under Information/System Information

- RetroArch: 0bba2d4

### Environment information

- OS: All OSes
- Compiler: [In case you are running local builds]

### PR DESCRIPTION
**Pull request by Jamiras**
## Description

Creates a wrapper object that contains the core state and achievement state and uses that for save states and rewind states. The wrapper is created before applying RZIP if enabled. Netplay and runahead still use the raw unwrapped core state.

The basic structure of the wrapper object is an 8 byte marker followed by one or more chunks. Each chunk has an 8 byte header identifying the chunk type and its size. This adds 32 bytes plus the achievement state size to each save state (before compression).

If the 8 byte marker is not seen when loading a state, the state is assumed to be the old format (raw core state) and passed directly to the core. New states cannot be loaded by the old code as the entire wrapper object would be passed to the core and the core would not accept it.

To address the fact that rewind uses a fixed state size, it must be stopped and restarted once the achievements are loaded to account for the additional space required for the achievement state data. As achievements are unlocked, the achievement state shrinks, but the rewind buffer size is not further modified. The wrapper object simply won't fill the entire allocated space.

## Related Issues

fixes #11894

## Related Pull Requests

n/a

## Reviewers

@twinaphex @Alcaro @Sanaki

### CODE REVIEW

**Review by Sanaki — APPROVED** [review]
I've been testing this on and off for a few days now and haven't found any functional issues. It's a damn nice feature, and one we've had requested more times than I can count. That said, I can't speak to any of the philosophical concepts behind what a state file should contain, if any exist.

File: tasks/task_save.c [hunk 1]
```diff
@@ -575,35 +593,124 @@ static void task_save_handler_finished(retro_task_t *task,
...
+#ifdef HAVE_CHEEVOS
+   size->cheevos_size = rcheevos_get_serialize_size();
+   if (size->cheevos_size > 0)
+      size->total_size += 8 + content_align_size(size->cheevos_size); /* 8-byte block header + content */
+#endif
+
+   return true;
+}
+
+size_t content_get_serialized_size()
```
Comment (twinaphex) [inline tasks/task_save.c:623]: Use (void) in function declarations/prototypes instead of ().

**Review by twinaphex — CHANGES_REQUESTED** [review]

File: tasks/task_save.c (hunk 1 above)
Comment (Jamiras) [reply tasks/task_save.c]: done

**Comment by sonninnos** [comment]
Unfortunately this breaks "Undo Load State".

**Comment by m4xw** [comment]
Hey @Jamiras , I would recommend we make this optional in some way, I would prefer if my savestates in mupen stay compatible with their standalone counterparts, I just became aware of this PR and I think this will likely fuck things up.
Also the state should only be ever appened to (i didnt go through the PR in detail yet), this is smth we should enforce.
Other programs just have a magic at the end, instead of the start, in such cases.

**Comment by Jamiras** [comment]
Wouldn't rzip (compressed save states) also "fuck things up"?

This PR makes the retroarch save state a container (similar to mp4). There's a versioned header, then labelled blocks. The program can handle the blocks it cares about as it encounters them and ignores the blocks that it doesn't understand. This format allows additional blocks in the future without breaking compatibility (i.e. screenshot, core/version information, etc).

If the versioned header is not present, then the old behavior is assumed and the full file is passed to the core.

> a core asserting that the sum of state parts is != state size could i see hapen too

The cores [do often do this](https://github.com/libretro/libretro-fceumm/blob/master/src/drivers/libretro/libretro.c#L1973-L1974). If the size passed to [retro_unserialize](https://github.com/libretro/RetroArch/blob/55c855267e68f97b58e1a0fa87c9aa3e70e3713c/libretro-common/include/libretro.h#L3425) is not correct, the data is discarded as corrupt or outdated. The container block knows the size of the data that was captured when it was created, so the correct subset of the file will be passed to the core.

Without a container or some other way of chunking the data, how do you know what portion of the file to pass to the core? You can't use `retro_serialize_size()`, because it's the current size and will always match the expectation in the unserialize function. Using it to split an old file for a newer version of the core eliminates the safeguard and may lead to unintended side effects.

**Comment by m4xw** [comment]
> Wouldn't rzip (compressed save states) also "fuck things up"?

They do, but its a option.

> This PR makes the retroarch save state a container (similar to mp4). There's a versioned header, then labelled blocks. The program can handle the blocks it cares about as it encounters them and ignores the blocks that it doesn't understand. This format allows additional blocks in the future without breaking compatibility (i.e. screenshot, core/version information, etc)

Dont get me wrong, I am for it. But we still need to be able to have core produce their states in a compatible way. This is especially important for debugging and migration for end users

> Without a container or some other way of chunking the data, how do you know what portion of the file to pass to the core? You can't use retro_serialize_size(), because it's the current size and will always match the expectation in the unserialize function. Using it to split an old file for a newer version of the core eliminates the safeguard and may lead to unintended side effects.

In internal chat I suggested a binary .meta file, which is just that. I am up for suggestions.
I also wouldnt say its urgent, but it def used to be a nice feature that i would like to retain.
At the very least all the header stuff should be moved to the end of the file.

**Comment by m4xw** [comment]
I just want to avoid cores having to do their own file io for this ultimately.

**Comment by m4xw** [comment]
> Without a container or some other way of chunking the data, how do you know what portion of the file to pass to the core? 

I want to add, kernel signing f.e. checks for a magic at the end of a module, that way it knows if eof data are available.
Then it makes a backwards sizeof() calc, copies/casts the header, verifies the data, passes the start of data to other things. That way existing tools dont have issues with such files, same use case at least.
But i really thing we need to seperate "frontend savestate" and "core savestate" for this. They shouldnt be the same file.

**Comment by Jamiras** [comment]
> I just want to avoid cores having to do their own file io for this ultimately.

Cores shouldn't have to do any I/O. The frontend gets the data from the save state and passes it to the core.

> i really think we need to seperate "frontend savestate" and "core savestate" for this

I don't think there is any sort of "core savestate". The frontend asks the core for data to write to a file and passes that same data back to the core at a later time. How the frontend manages that data is up to the frontend. All save states are "frontend savestates". Multiple frontends may support the "naked" core state data, but it doesn't make it a "core savestate".

> At the very least all the header stuff should be moved to the end of the file.

Prior to this PR, the entire file was assumed to be the core's state data. The file contents and file size were passed to `retro_unserialize` without any additional processing. As soon as we put anything else in the file it can no longer be passed directly to the core because the size won't be correct, so why does it matter where the extra data goes? 

> Savestate file output is implementation defined

> This also breaks copying libretro states between different libretro frontends, which is absolutely supported.

If the save state file is implementation defined, how can they be copied between frontends?

> This is especially important for debugging and migration for end users

I can see the desire to share state information for debugging purposes, but migration is probably better handled by SRAM (which is also just raw data provided by the core). I wouldn't expect to be able to use Snes9x save states in BSnes or RetroArch, but I would expect to be able to copy my in-game saves between them as the cartridge RAM should not differ between emulators.

Perhaps the rzip tool (https://github.com/libretro/RetroArch/tree/master/libretro-common/samples/streams/rzip) which is already provided for rzipping and un-rzipping save states could be extended to also support extracting the raw core state.

**Pull request merged**